### PR TITLE
Fix intergrating hyper-parameters

### PR DESCRIPTION
**Pull request by marpulli**
Integrating hyper-parameters didn't work with multi-output models or acquisitions without gradients.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/core/acquisition/integrated_acquisition.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 
 from emukit.core.acquisition import Acquisition
-from emukit.core.interfaces import IModel, IPriorHyperparameters, IDifferentiable
```
Comment (apaleyes) [inline emukit/core/acquisition/integrated_acquisition.py]: Isn't it actually used? I mean, there is definitely a call of `model.evaluate_with_gradients` below. I'd suggest to add this interface to model type hint

File: emukit/core/acquisition/integrated_acquisition.py [hunk 2]
```diff
@@ -47,18 +50,21 @@ def evaluate_with_gradients(self, x: np.ndarray) -> Tuple:
...
 
-        return acquisition_value / self.n_samples, dacquisition_dx / self.n_samples
+    def update_parameters(self):
+        self.samples = self.model.generate_hyperparameters_samples(self.n_samples)
 
     @property
     def has_gradients(self) -> bool:
         """Returns that this acquisition has gradients"""
-        return isinstance(self.model, IDifferentiable)
+        return self._has_gradients
```
Comment (apaleyes) [inline emukit/core/acquisition/integrated_acquisition.py]: Let's cover this with a unit test

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 3]
```diff
@@ -241,3 +240,33 @@ def get_covariance_between_points(self, X1: np.ndarray, X2: np.ndarray) -> np.nd
...
+        samples = hmc.sample(num_samples = n_burnin + n_samples * subsample_interval, hmc_iters = leapfrog_steps)
+        hmc_samples = samples[n_burnin::subsample_interval]
+        return hmc_samples
+
+    def fix_model_hyperparameters(self, sample_hyperparameters: np.ndarray) -> None:
+        """
+        Fix model hyperparameters
+
+        """
+        if self.gpy_model._fixes_ is None:
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py]: Is there a way in gpy to do this without calling private stuff like `_fixes_` and `_trigger_params_changed`?

File: emukit/core/acquisition/integrated_acquisition.py (hunk 1 above)
Comment (marpulli) [reply emukit/core/acquisition/integrated_acquisition.py]: I have always been confused how we should do type hints when some interfaces are optional to implement. The type hint should say "Must implement IModel and optionally implement IDifferentiable". We usually do `Union[IModel, IDifferentialbe]` which means "must implement both IModel and IDifferentialbe".  You get warnings from the type checking system if you pass in a model that doesn't implement gradients with the current type hints.

File: emukit/core/acquisition/integrated_acquisition.py (hunk 2 above)
Comment (marpulli) [reply emukit/core/acquisition/integrated_acquisition.py]: Done

File: emukit/model_wrappers/gpy_model_wrappers.py (hunk 3 above)
Comment (marpulli) [reply emukit/model_wrappers/gpy_model_wrappers.py]: I'll investigate, this is the way the other model wrapper does it though

Comment (apaleyes) [reply emukit/model_wrappers/gpy_model_wrappers.py]: Following our offline conversation, can you please create an issue to remind us to look deeper into this GPy magic? We may even consider contributing interface changes to GPy if necessary.

**Comment by apaleyes** [comment]
This got broken after #241, and I've submitted a PR with a fix already: #244

**Pull request merged**